PrebidMobile/PrebidMobileCore/src/main/java/org/prebid/mobile/core/Prebid.java

```diff
@@ -236,12 +287,53 @@ private static void handleDFPCustomTargetingUpdate(Object adRequestObj, String a
         if (bundle != null) {
             ArrayList<Pair<String, String>> prebidKeywords = BidManager.getKeywordsForAdUnit(adUnitCode, context);
             if (prebidKeywords != null && !prebidKeywords.isEmpty()) {
-                // retrieve keywords from mopub adview
-                for (Pair<String, String> keywordPair : prebidKeywords) {
-                    bundle.putString(keywordPair.first, keywordPair.second);
-                    usedKeywordKeys.add(keywordPair.first);
+                boolean bidShouldBeLoadedInDemandSdk = false;
+                String bidder = "";
+                for (Pair<String, String> p : prebidKeywords) {
+                    if (PREBID_CREATIVE_LOAD_TYPE.equals(p.first)) {
+                        if (PREBID_DEMAND_SDK.equals(p.second)) {
+                            bidShouldBeLoadedInDemandSdk = true;
+                        }
+                    }
+                    if (PREBID_BIDDER.equals(p.first)) {
+                        bidder = p.second;
+                    }
+                }
+                if (bidShouldBeLoadedInDemandSdk) {
+                    // pass bids that require demand sdk only when rendering is enabled
+                    // save the cache id locally on the ad object for later use
+                    Class mediation_adapter = null;
+                    AdUnit adUnit = BidManager.getAdUnitByCode(adUnitCode);
+                    if (AdType.BANNER.equals(adUnit.getAdType())) {
+                        mediation_adapter = getClassFromString(PREBID_DFP_CUSTOM_EVENT_BANNER);
+                    } else if (AdType.INTERSTITIAL.equals(adUnit.getAdType())) {
+                        mediation_adapter = getClassFromString(PREBID_DFP_CUSTOM_EVENT_INTERSTITIAL);
+                    }
+                    if (mediation_adapter != null && PrebidDemandSettings.isDemandEnabled(bidder)) {
+                        for (Pair<String, String> keywordPair : prebidKeywords) {
+                            bundle.putString(keywordPair.first, keywordPair.second);
+                            usedKeywordKeys.add(keywordPair.first);
+                            // set custom event extras for the requested type
+                            if (mediation_adapter != null) {
+                                Bundle customEventExtras = (Bundle) callMethodOnObject(adRequestObj, "getCustomEventExtrasBundle", mediation_adapter);
```

Comment:
Again, if this doesn't need to be done in the loop, better to loop first and get cacheId and bidder and then set it on DFP customEventExtras.